CONTRIBUTING.md

```diff
@@ -56,14 +36,129 @@ When creating your enhancement request, please:
+    
+        git checkout master  # if not already on master
+        git pull --rebase upstream master
+        git push
+
+    </details>
+
+3. **Create a new branch to work on your fix or enhancement**
+
+    <details><summary>Expand details 👇</summary><br/>
 
-We are always looking for new models to add to our collection.  If you have trained a model and would like to include it in
-AllenNLP, please create [a pull request](https://github.com/allenai/allennlp/pulls) that includes:
+    Commiting directly to the master branch of your fork is not recommended. It will be easier to keep your fork clean if you work on a seperate branch for each contribution you intend to make.
+    
+    You can create a new branch with
+    
+        # replace BRANCH with whatever name you want to give it
+        git checkout -b BRANCH
+        git push --set-upstream origin BRANCH
+
+    </details>
+
+4. **Test your changes**
+
+    <details><summary>Expand details 👇</summary><br/>
+
+    Our continuous integration (CI) testing runs [a number of checks](https://github.com/allenai/allennlp/actions?query=workflow%3APR) for each pull request on [GitHub Actions](https://github.com/features/actions). You can run most of these tests locally, which is something you should do *before* opening a PR to help speed up the review process and make it easier for us.
+    
+    First, you should run [`black`](https://github.com/psf/black) to make sure you code is formatted consistently. Many IDEs support code formatters as plugins, so you may be able to setup black to run automatically everytime you save. [`black.vim`](https://github.com/psf/black/tree/master/plugin) will give you this functionality in Vim, for example. But `black` is also easy to run directly from the command line. Just run this from the root of your clone:
+    
+        black .
+
+    Our CI also uses [`flake8`](https://github.com/allenai/allennlp/tree/master/tests) to lint the code base and [`mypy`](http://mypy-lang.org/) for type-checking. You should run both of these next with
+
+        make lint
+
+    and
+
+        make typecheck
+
+    We also strive to maintain high test coverage, so most contributions should include additions to [the unit tests](https://github.com/allenai/allennlp/tree/master/tests). These tests are ran with [`pytest`](https://docs.pytest.org/en/latest/), which you can use to locally run any test modules that you've added or changed.
+
+    For example, if you've fixed a bug in `allennlp/nn/util.py`, you can run the tests specific to that module with
+    
+        pytest -v tests/nn/util_test.py
+    
+    Our CI will automatically check that test coverage stays above a certain threshold (around 90%). To check the coverage locally in this example, you could run
+    
+        pytest -v --cov allennlp.nn.util tests/nn/util_test.py
+
+    You should also make sure that the API documentation can build without errors. For that, just run
```

Comment:
Maybe some note here that failures here are pretty rare, and you probably won't have to run this.  But if you need to, you can run it locally with this.